src/lib.rs

```diff
@@ -96,75 +101,79 @@ impl<'a> MownStr<'a> {
         debug_assert!(self.is_owned(), "extract_box() called on borrowed MownStr");
         // extract data to make box
         let ptr = self.addr.as_ptr();
-        let len = self.real_len();
         // turn to borrowed, to avoid double-free
-        self.xlen = 0;
+        self.owners = None;
         debug_assert!(self.is_borrowed());
         // make box
-        let slice = slice::from_raw_parts_mut(ptr, len);
+        let slice = slice::from_raw_parts_mut(ptr, self.len);
         let raw = str::from_utf8_unchecked_mut(slice) as *mut str;
         Box::from_raw(raw)
     }
 }
 
 impl<'a> Drop for MownStr<'a> {
     fn drop(&mut self) {
-        if self.is_owned() {
+        self.owners
+            .as_ref()
+            .map(|o| o.fetch_sub(1, Ordering::Relaxed));
+        if self.cleanup() {
```

Comment:
Can this be made to perform one atomic operation instead of two, and then keep the equivalence check still being `==`?